code/game/machinery/computer/dropship_weapons.dm

```diff
@@ -438,189 +673,464 @@
+		to_chat(weapon_operator, SPAN_WARNING("You don't have the training to fire this weapon!"))
+		return FALSE
+
+	return TRUE
+
+/obj/structure/machinery/computer/dropship_weapons/proc/ui_open_fire(obj/docking_port/mobile/marine_dropship/dropship, targ_id)
+	var/mob/weapon_operator = usr
+	if(ishuman(weapon_operator))
+		var/mob/living/carbon/human/human_operator = weapon_operator
+		if(!human_operator.allow_gun_usage)
+			to_chat(human_operator, SPAN_WARNING("Your programming prevents you from operating dropship weaponry!"))
+			return FALSE
+	var/obj/structure/dropship_equipment/weapon/DEW = selected_equipment
+	if(!selected_equipment || !selected_equipment.is_weapon)
+		to_chat(weapon_operator, SPAN_WARNING("No weapon selected."))
+		return FALSE
+	if(!skillcheck(weapon_operator, SKILL_PILOT, DEW.skill_required)) //only pilots can fire dropship weapons.
+		to_chat(weapon_operator, SPAN_WARNING("You don't have the training to fire this weapon!"))
+		return FALSE
 
-	if(href_list["cas_camera"])
-		if(!ishuman(usr))
-			to_chat(usr, SPAN_WARNING("You have no idea how to do that!"))
-			return
-		if(dropship.mode != SHUTTLE_CALL)
-			to_chat(usr, SPAN_WARNING("Shuttle has to be in orbit."))
-			return
+	if(!faction)
+		return FALSE//no faction, no weapons
 
-		if(!faction)
-			to_chat(usr, SPAN_DANGER("Bug encountered, this console doesn't have a faction set, report this to a coder!"))
-			return
+	var/datum/cas_iff_group/cas_group = cas_groups[faction]
 
-		var/datum/cas_iff_group/cas_group = cas_groups[faction]
-		if(!cas_group)
-			to_chat(usr, SPAN_DANGER("Bug encountered, no CAS group exists for this console, report this to a coder!"))
-			return
+	if(!cas_group)
+		return FALSE//broken group. No fighting
+
+	for(var/datum/cas_signal/LT in cas_group.cas_signals)
+		if(LT.target_id == targ_id && LT.valid_signal())
+			if(dropship.mode != SHUTTLE_CALL)
+				to_chat(weapon_operator, SPAN_WARNING("Dropship can only fire while in flight."))
+				return FALSE
+			if(dropship.door_override)
+				return FALSE
+			if(!selected_equipment || !selected_equipment.is_weapon)
+				to_chat(weapon_operator, SPAN_WARNING("No weapon selected."))
+				return FALSE
+			DEW = selected_equipment // for if the weapon somehow changes
+			if(!skillcheck(weapon_operator, SKILL_PILOT, DEW.skill_required)) //only pilots can fire dropship weapons.
+				to_chat(weapon_operator, SPAN_WARNING("You don't have the training to fire this weapon!"))
+				return FALSE
+			if(!dropship.in_flyby && DEW.fire_mission_only)
+				to_chat(weapon_operator, SPAN_WARNING("[DEW] requires a fire mission flight type to be fired."))
+				return FALSE
+
+			if(!DEW.ammo_equipped || DEW.ammo_equipped.ammo_count <= 0)
+				to_chat(weapon_operator, SPAN_WARNING("[DEW] has no ammo."))
+				return FALSE
+			if(DEW.last_fired > world.time - DEW.firing_delay)
+				to_chat(weapon_operator, SPAN_WARNING("[DEW] just fired, wait for it to cool down."))
```

Comment:
All of this can be taken out of for loop and arguably would be more helpful errors that way